### DISCUSSION TITLE
@task.virtualenv

### DISCUSSION THREAD

**Discussion by stefanondisponibile** [apache/airflow#19488] [Q&A] [answered]
I am a tad confused by the `@task.virtualenv` usage with the TaskFlow style.

How do you pass `templates_dict` in?
What's the actual use when you can pass `op_args` and `op_kwargs`, which are templated?

Thank you!

**Comment by potiuk** [comment, marked as answer]
Can you be more specific and give some examples of what you want to do ? I think the problem is that your question misses the context.

**Reply by stefanondisponibile** [reply]
Yes, sorry, I was probably missing some more context.
However, I think I figured it out, yet it looks a bit confusing to me:

```python
@task.virtualenv(
      use_dill=False,
      system_site_packages=False,
      requirements=['funcsigs'],
  )
def some_fn(**kwargs):
    print(kwargs)
    return None
task_x = some_fn(
    op_args=["foo"],
    op_kwargs={"task_id": "{{ task.task_id }}"},
    templates_dict={"dag_id": "{{ dag.dag_id }}"}
)
```

If I'm using this the right way, what I found confusing was the ability to pass arguments to both `some_fn`  _and_ the `task.virtualenv` decorator.

**Reply by potiuk** [reply]
The 'virtualenv' decorator parameters are the "configuration of the PythonVirtualenv operator under the hood. Where parameters in some_fn are the runtime parameters. Actually there is no need to pass the parameters this way.

I am not sure if you realise but you can specify all the "context" parameters you want to get in your decorated function and they will be automatically injected for you. There is no need ot use jinja templating for the decorated parameters - either directly or via kwargs dict. For example to access "task instance" from the context you can do  this:

```
def some_fn(ti=None, **kwargs)
    # use ti here
```

or 

```
def some_fn(**kwargs)
   ti = kwargs['ti'] 
``

**Reply by josh-fell** [reply]
To add to @potiuk's answer, more generally you don't need to specify `op_args` and `op_kwargs` when using the TaskFlow API. You can use the normal Python syntax for passing in args to a function:
```python
@task
def some_func(input1, input2, ti=None):
    ...

task_x = some_func(input1="my_arg", input2="my_other_arg")
```

**Reply by stefanondisponibile** [reply]
Thank you @potiuk  and thank you @josh-fell !
